components/calendar/Header.tsx

```diff
@@ -139,26 +180,69 @@ export default class Header extends React.Component<HeaderProps, any> {
     this.calenderHeaderNode = node;
   };
 
-  renderHeader = ({ getPrefixCls }: ConfigConsumerProps) => {
-    const { prefixCls: customizePrefixCls, type, value, locale, fullscreen } = this.props;
+  getMonthYearSelections = (getPrefixCls: ConfigConsumerProps['getPrefixCls']) => {
+    const { prefixCls: customizePrefixCls, type, value } = this.props;
+
     const prefixCls = getPrefixCls('fullcalendar', customizePrefixCls);
-    const yearSelect = this.getYearSelectElement(prefixCls, value.year());
-    const monthSelect =
+    const { yearReactNode, yearList } = this.getYears(prefixCls, value.year());
+    const { monthReactNode, monthList } =
       type === 'month'
-        ? this.getMonthSelectElement(prefixCls, value.month(), this.getMonthsLocale(value))
-        : null;
+        ? this.getMonths(prefixCls, value.month(), this.getMonthsLocale(value))
+        : { monthList: [], monthReactNode: null };
+    return {
+      yearReactNode,
+      yearList,
+      monthReactNode,
+      monthList,
+    };
+  };
+
+  getTypeSwitch = () => {
+    const { locale, type, fullscreen } = this.props;
     const size = fullscreen ? 'default' : 'small';
-    const typeSwitch = (
+    return (
       <Group onChange={this.onTypeChange} value={type} size={size}>
         <Button value="month">{locale.month}</Button>
         <Button value="year">{locale.year}</Button>
       </Group>
     );
+  };
+
+  headerRenderCustom = (getPrefixCls: ConfigConsumerProps['getPrefixCls']): React.ReactNode => {
+    const { headerRender, type } = this.props;
+    const typeSwitch = this.getTypeSwitch();
+    const { yearReactNode, monthReactNode, monthList, yearList } = this.getMonthYearSelections(
+      getPrefixCls,
+    );
+
+    const callBackData: RenderHeader = {
+      type,
+      typeSwitch,
+      typeChange: this.onTypeChange,
+      years: yearList,
+      yearSelect: yearReactNode,
+      yearChange: this.onYearChange,
+      months: monthList,
+      monthSelect: monthReactNode,
+      monthChange: this.onMonthChange,
+      ref: this.getCalenderHeaderNode,
+    };
```

Comment:
How about just pass args:
```tsx
interface HeaderRenderArgs {
  value: Moment;
  onChange: (value: Moment) => void;
  type: string;
  onTypeChange: (type: string) => void;
}
```

Since if we export to user, args should be simplify.

Reply:
@zombieJ your example very clear i aggri with you.

But in this case I don't have single value and single change event 

for render custom header I think we need to more arguments now I pass arguments for
` yearSelect` - is ant select box with option I can remove this option this is for more flexibility
`monthSelect` - some here
`typeSwitch` - I'm using some switch node because now we are supporting only year and month I think I will add that functional in future
`typeChange` this is for customize type change event but for now I can remove these arguments too
`monthChange` I'm using some function for the change  month
`yearChange` some here 
 `months` is months array this is for more customize I think we can have some cases when engineer want to use other select box in that case we will just use this array 
`years`:  some here 
`type` is selected type `month|year` but in the future I think we can have more types 
`ref` is select ref I can remove these arguments too



with my argument you can change select box or just positions, please take a look

this is for just change input ordering
```js
headerRender = {
  ({ yearSelect,  monthSelect, typeSwitch}) => (
      <Row>
         {yearSelect}
         {typeSwitch}
         {monthSelect}
     </Row>
)
}
```

also, we can use more custom for example 

```js
headerRender = {
  ({ months,  years, typeSwitch, monthChange, yearChange}) => (
      <div style={{ padding: 10, border: '1px solid #d9d9d9' }} ref={returnData.ref}>
          <Select defaultValue="Mar" onChange={monthChange} style={{ width: '100px' }}>
            {months.map(item => (
              <Select.Option key={`${item.index}`}>{item.month}</Select.Option>
            ))}
          </Select>
          <Select onChange={yearChange} style={{ width: '100px' }}>
            {years.map(item => (
              <Select.Option key={`${item.index}`}>{item.index}</Select.Option>
            ))}
          </Select>
          typeSwitch
        </div>
)
}

```

I think this example more flexible for the future for engeners. What do you think @zombieJ?

Reply:
Hi @abgaryanharutyun, `yearSelect` & `MonthSelect` is nothing special. User can generate same ReactNode with just current moment `value`(instead of row year & month data or nodes).

```jsx
// Just sample, not real code

const year = value.years();

function onYearChange(newYear) {
  const now = value.clone().years(newYear);
  onChange(value);
}

const options = [];
for (let i = year - 100; i < year + 100; i += 1) {
  options.push(
    <Option key={i} value={i}>{i}</Option>
  );
}

const customizeYearSelect = (
  <Select
    size="small"
    dropdownMatchSelectWidth={false}
    className="my-year-select"
    onChange={onYearChange}
    value={String(year)}
    getPopupContainer={() => someFuncAnyway}
>
  {options}
</Select>
```

When pass the customize ability to user, we should follow Ockham's Razor to avoid pass too many internal data/nodes to user. Since current arguments is not necessary.
These 4 arguments can already do what need to do:
```tsx
interface HeaderRenderArgs {
  value: Moment; // yearSelect, monthSelect, months, years
  onChange: (value: Moment) => void; // monthChange, yearChange
  type: string; // type, typeSwitch
  onTypeChange: (type: string) => void; // typeChange
}

// ref is just let user wrapped as customize div instead:
headerRender() {
  return (
    <div ref={customizeRef}>
      ...
    </div>
  );
}
```

Reply:
Got it thanks @zombieJ I will fix

Reply:
@zombieJ you can take a look thanks for you good point.